### PR TITLE
Implement draggable shapes

### PR DESCRIPTION
**Pull request by n-riesco**
* Added drag element to shapes
* Added tests
* Addresses https://github.com/plotly/streambed/issues/1525#issuecomment-225043702

---

Resubmitted PR after:
- removing `console.log` from tests,
- and checking tests pass locally with both Chromium and Firefox.

### CODE REVIEW

**Comment by n-riesco** [comment]
I've run the tests locally using the CircleCI configuration (`node_modules/.bin/karma start test/jasmine/karma.ciconf.js "tests/shapes_test.js"`) and they pass.

Any ideas how I can reproduce the failures locally?

**Comment by n-riesco** [comment]
I can reproduce the issue locally by using Firefox 38.0.

**Comment by n-riesco** [comment]
I've fixed the failures in CircleCI!!!
The PR is ready for review.

File: test/jasmine/tests/shapes_test.js [hunk 1]
```diff
@@ -414,9 +414,9 @@ describe('Test shapes', function() {
     }
 
     function getShapeNode(index) {
-        return d3.selectAll('.shapelayer')
-            .selectAll('path[data-index="' + index + '"')
-            .node();
+        return d3.selectAll('.shapelayer path').filter(function() {
+            return +this.getAttribute('data-index') === index;
```
Comment (etpinard) [inline test/jasmine/tests/shapes_test.js]: Interesting. Nice find @n-riesco !

File: src/components/shapes/index.js [hunk 2]
```diff
@@ -343,6 +345,72 @@ function updateShape(gd, index, opt, value) {
...
                 'clip' + gd._fullLayout._uid + clipAxes);
         }
+
+        if(!gd._context.editable) return;
+
+        var update;
+        var x0, y0, x1, y1, astrX0, astrY0, astrX1, astrY1;
+        var pathIn, astrPath;
+        var xa, ya, x2p, y2p, p2x, p2y;
+        dragElement.init({
```
Comment (etpinard) [inline src/components/shapes/index.js]: Nice :recycle: action !

File: test/jasmine/tests/shapes_test.js [hunk 3]
```diff
@@ -282,3 +285,399 @@ describe('Test shapes: a plot with shapes and an overlaid axis', function() {
...
+    'use strict';
+
+    var gd, data, layout, config;
+
+    beforeEach(function() {
+        gd = createGraphDiv();
+        data = [{}];
+        layout = {};
+        config = {
+            editable: true,
```
Comment (etpinard) [inline test/jasmine/tests/shapes_test.js]: Yep. This is the way to go. The other `editable: true` are currently in [`config_test.js`](https://github.com/plotly/plotly.js/blob/3c85d9eb47ce08177d951065fd66a7d9f9671e57/test/jasmine/tests/config_test.js#L45), but scoping them by the component(s) they affect is much better. Thanks!

**Comment by n-riesco** [comment]
# TODO
- [ ] test shape resize
- [ ] update cursor onmousemove

File: src/components/shapes/index.js [hunk 4]
```diff
@@ -346,74 +346,79 @@ function updateShape(gd, index, opt, value) {
...
-                if(dragged) {
-                    Plotly.relayout(gd, update);
-                }
-            }
-        });
+        if(gd._context.editable) setupDragElement(gd, path, options, index);
     }
 }
 
+function setupDragElement(gd, shapePath, shapeOptions, index) {
```
Comment (etpinard) [inline src/components/shapes/index.js]: I love this. Way to flatten the code. Thanks!

File: src/components/shapes/index.js [hunk 5]
```diff
@@ -343,6 +345,157 @@ function updateShape(gd, index, opt, value) {
...
+            x1 = x2p(shapeOptions.x1);
+            y1 = y2p(shapeOptions.y1);
+
+            astrX0 = astr + '.x0';
+            astrY0 = astr + '.y0';
+            astrX1 = astr + '.x1';
+            astrY1 = astr + '.y1';
+        }
+
+        if(x0 < x1) {
```
Comment (etpinard) [inline src/components/shapes/index.js]: Great. Even better than adding multiple drag elements. :tada:

**Comment by n-riesco** [comment]
I've been thinking of an algorithm to resize path shapes. Here's a simple proposal:
- scale the x coordinate by a factor of `1 + dx/element.getBoundingClientRect.width`
- scale the y coordinate by a factor of `1 + dy/element.getBoundingClientRect.height`
- limit the scale down `element.getBoundingClientRect.width - dx > MINWIDTH`
- limit the scale down `element.getBoundingClientRect.height - dy > MINHEIGHT`

File: src/components/shapes/index.js [hunk 6]
```diff
@@ -361,17 +361,35 @@ function setupDragElement(gd, shapePath, shapeOptions, index) {
...
-    function startDrag(evt, startX, startY) {
+    function updateDragMode(evt) {
+        // choose 'move' or 'resize'
+        // based on initial position of cursor within the drag element
+        var w = dragBBox.right - dragBBox.left,
+            h = dragBBox.bottom - dragBBox.top,
+            x = evt.clientX - dragBBox.left,
+            y = evt.clientY - dragBBox.top,
+            cursor = (w > MINWIDTH && h > MINHEIGHT) ?
+                dragElement.getCursor(x / w, 1 - y / h) :
```
Comment (etpinard) [inline src/components/shapes/index.js]: Nicely done here.

File: src/components/shapes/index.js [hunk 7]
```diff
@@ -379,7 +379,7 @@ function setupDragElement(gd, shapePath, shapeOptions, index) {
             h = dragBBox.bottom - dragBBox.top,
             x = evt.clientX - dragBBox.left,
             y = evt.clientY - dragBBox.top,
-            cursor = (w > MINWIDTH && h > MINHEIGHT) ?
+            cursor = (w > MINWIDTH && h > MINHEIGHT && !evt.shiftKey) ?
```
Comment (etpinard) [inline src/components/shapes/index.js]: good.

File: src/components/shapes/index.js [hunk 8]
```diff
@@ -343,6 +345,160 @@ function updateShape(gd, index, opt, value) {
...
+            element: shapePath.node(),
+            prepFn: startDrag,
+            doneFn: endDrag
+        },
+        dragBBox = dragOptions.element.getBoundingClientRect(),
+        dragMode;
+
+    dragElement.init(dragOptions);
+
+    function updateDragMode(evt) {
```
Comment (etpinard) [inline src/components/shapes/index.js]: Great. We should add a `setCursor` handler for annotations too.

That will deserve a PR of it own. Ticket here: https://github.com/plotly/plotly.js/issues/679

File: test/jasmine/tests/shapes_test.js [hunk 9]
```diff
@@ -282,3 +285,518 @@ describe('Test shapes: a plot with shapes and an overlaid axis', function() {
...
+        {
+            title: 'linked to linear and log axes',
+            xaxis: { type: 'linear', range: [0, 10] },
+            yaxis: { type: 'log', range: [Math.log10(1), Math.log10(1000)] }
+        },
+
+        // xaxis.type: 'date', yaxis.type: 'category'
+        {
+            title: 'linked to date and category axes',
+            xaxis: { type: 'date', range: ['2000-01-01', '2000-01-02'] },
```
Comment (etpinard) [inline test/jasmine/tests/shapes_test.js]: Hmm. `range` for date axes currently only work when set to timestamps. See example: http://codepen.io/etpinard/pen/RRoQKr

**Comment by etpinard** [comment]
@n-riesco shapes on referenced to date axes currently support date string (e.g '2016-06-22') and timestamps (e.g `(new Date(2016, 5, 22).getTime()`). See example: http://codepen.io/etpinard/pen/oLYEGN

But with `{ editable: true }` dragging the timestamp shape results in 

image

on line

image

**Comment by n-riesco** [comment]
@etpinard Thx for catching that one! I've just pushed a fix and updated the tests to use both date strings and timestamps.

File: src/components/shapes/index.js [hunk 10]
```diff
@@ -528,7 +528,10 @@ function isShapeInSubplot(gd, shape, plotinfo) {
 }
 
 function decodeDate(convertToPx) {
-    return function(v) { return convertToPx(v.replace('_', ' ')); };
+    return function(v) {
+        if(v.replace) v = v.replace('_', ' ');
+        return convertToPx(v);
```
Comment (etpinard) [inline src/components/shapes/index.js]: nice

**Comment by etpinard** [comment]
Great PR. Thanks @n-riesco :beers:

File: src/components/shapes/index.js [hunk 11]
```diff
@@ -346,74 +346,79 @@ function updateShape(gd, index, opt, value) {
...
+                astrX1 = astr + '.x1';
+                astrY1 = astr + '.y1';
+            }
+
+            update = {};
+        },
+        moveFn: function(dx, dy) {
+            if(shapeOptions.type === 'path') {
+                var moveX = function moveX(x) { return p2x(x2p(x) + dx); };
+                if(xa && xa.type === 'date') moveX = encodeDate(moveX);
```
Comment (yogat3ch) [inline src/components/shapes/index.js]: Hi @etpinard,
Is it possible to [constrain the drag direction](https://api.jqueryui.com/draggable/#option-axis) to the x or y axis based on an option value?

**Pull request merged**